spec/fixtures/router_path_handling_tests.lua

```diff
@@ -0,0 +1,194 @@
+-- The following tests are used by unit and integration tests
+-- to test the router path handling. Putting them here avoids
+-- copy-pasting them in several places.
+--
+-- The tests can obtain this table by requiring
+-- "spec.fixtures.router_path_handling_tests"
+--
+-- Note that the tests are parsed into a hash form at the end
+-- of this file before they are returned
+--
+-- All the tests where v1 differs from v0 are marked with !
+
+local tests = {
+  -- upstream_path   route_path   strip_path  path_handling  request_path    expected_path           --    0
```

Comment:
In general I think this is easier to read if it is:
```
-- request_path     route_path     service_path     expected_path     strip_path
```

`strip_path` can be on either side I think.

Reply:
Why do you think that is better? I decided to go with "Service attributes / Route Attributes / Request / Result" because that's the order in which one would set up things when doing the tests manually: first create the Service, then the Route, then make a request, and then check the resulting path.

Reply:
I am thinking it in terms of request:

1. there is request path (that comes in)
2. that path is matched to route path (and route has an service)
3. request is proxied to service_path
4. upstream path is constructed from service path and request path

Reply:
@kikito please check with @thibaultcha as he said that it is impossible to read.

I think originally when this was created was thinking a bit from `concatenation` perspective, but stripping makes you jump back and forth.

in that case I think the best order would be:
```
route_path  service_path  request_path  proxied_path
```

`service_path` is concatenated with `request_path` ... but that can be stripped by `route_path`. So... I think it is best to order it chronologically on what happens when request comes in.

Reply:
I see now, thanks for explaining.

You see the test from the perspective of "a request coming to an existing system, and what it *encounters*". I see it from the POV of "a programmer building tests step by step".

> stripping makes you jump back and forth

I believe putting `strip_path` or `path_handling` at the end makes you jump back and forth too. I believe that they belong next to `route_path`, as they are part of the Route object, and they are "encountered" before we can get a "result" (`proxied_path`).

However I still believe the order I chose is more understandable because it is similar to what we do in most tests. That is the strongest point about it IMHO.

Reply:
I have no idea how the routes and services are build. There is no actual order. You can create them in almost any order. Thus hard to wrap in minde "a programmer building tests". This is a matrix of tests that you read from top to bottom and the easiest way is to memorize them as they enter to system, and not by some arbitrary order some specific programmer decided to follow.

Reply:
Yes, `route_strip_path` and `route_path_handling` can be specified next to `route_path`.

Reply:
```
-- request_path  route_path  route_stip_path  route_path_handling  service_path  proxied_path  strip_path
```

As your order is almost the same as before and it was voiced several times that it is hard to read, we probably want something else.

Reply:
> There is no actual order. You can create them in almost any order [...] some arbitrary order some specific programmer decided to follow.

I can understand that it is not what you would have preferred, but it is hardly arbitrary. You can not create a Route without creating the Service first. And you can't make a (successful) request without creating a Route first. And you can't check a result without doing the request first.

We have used this order before many times in other tests. I honestly don't know how to explain this better.

> It was voiced several times that it is hard to read, we probably want something else.

The order for which it was voiced that it was hard to read had the `strip_path` at the end, separated from the rest of the Route attributes.

> This is a matrix of tests that you read from top to bottom and the easiest way is to memorize them as they enter to system, and not by some arbitrary order some specific programmer decided to follow.

I don't think it needs to be read from top to bottom; I its main purpose is to *facilitate finding the individual case you are looking for* (for debugging or adding new cases in the right spot). I also don't think memorizing the whole list is necessary, just the individual row one is studying. On that direction, here's some other changes we could make:

* Reorder the *rows* as well, so that the vertical order mimics the column order.
* Using more useful path names. Instead of `/fee/bor` we could use `/true/v0` for the `strip_path = true, path_handling = v0` cases, for example.

I didn't do those changes here because waddling through the git diffs is already quite daunting as they are now. If you think this could help, I can make those changes in an extra commit.

Reply:
I have added further commits refactoring the tests:
* Used more meaningful path names (i.e. `/service` instead of `/fee/bor`)
* Compressed the tests(if the same test works the same in `v0` and `v1`, it gets a single line named `{v0,v1}` instead of two.
* Reordered the tests so it's always `service attrs / route attrs / request`.